### PR TITLE
Addition of Depth-wise Separable Convolution (Separable Convolution) Layer. #2166

### PR DESCRIPTION
**Pull request by kartikdutt18**
Hi everyone, This is a draft PR. In Separable Convolution when groups = InSize it will perform DepthWise Convolution, else you can use to perform Grouped Convolution. This implementation is inspired from pytorch. In DepthWise Convolution each input channel is convolved separately, here number of groups will determine number of groups in which  input channels will be divided, if numGroups = 2 then input channels will be divided in two and those groups will be convolved separately.
To perform depth-wise separable convolution use outsize = insize and number of groups = insize.
The following tasks have been completed:

- [x] Complete Forward Propagation.
- [x] Declare Base Class for Separable Convolution.
- [x] Completed Serialization and InitializeSamePadding Function.
- [x]  Complete Backward function.
- [x]  Complete Gradient function.

To Do:
- [ ]  Complete Simple Forward Propagation test. (Grouped)
- [ ]  Complete Complex Forward Propagation test. (DepthWise)
- [x]  Complete tests for simple Convolution.
- [ ]  Complete Test for Backward, Gradient, Delta function.

### CODE REVIEW

**Comment by kartikdutt18** [comment]
Sorry, it took me so long to get this much done, I have my internal examination this week so I am caught in its preparation.

**Comment by zoq** [comment]
No worries, best of luck with your exams.

**Comment by kartikdutt18** [comment]
Thank a lot.

**Comment by kartikdutt18** [comment]
Hi @zoq, I needed your help with testing. I ran the following code in [python](https://colab.research.google.com/drive/1VSHDd7Tsg5TrCwR4vZXZ361OjYGOSWGE) and got sum as 113 but Convolution Layer as well as Seperable-Convolution-Layer gave same result i.e. 111.0 .
Thanks.

File: src/mlpack/tests/ann_layer_test.cpp [hunk 1]
```diff
@@ -2931,4 +2932,32 @@ BOOST_AUTO_TEST_CASE(ConvolutionLayerPaddingTest)
   module2.Backward(std::move(input), std::move(output), std::move(delta));
 }
 
+/**
+ * Simple Separable Convolution layer test.
+ */
+BOOST_AUTO_TEST_CASE(SimpleSeparableConvolutionLayerTest)
+{
+  arma::mat output, input, delta;
+  SeparableConvolution<> module1(1, 4, 2, 2, 1, 1, 0, 0, 3, 3, 1, "valid");
```
Comment (kartikdutt18) [inline src/mlpack/tests/ann_layer_test.cpp]: Hi @zoq, I needed your help while implementing tests for this layer.
I ran the following test in [python](https://colab.research.google.com/drive/1VSHDd7Tsg5TrCwR4vZXZ361OjYGOSWGE#scrollTo=zoQfUhjep2cg) and got sum as 111 but when I run the same test using Convolution Layer I get 113. Could you help me in figuring out what I did wrong.
Thanks.

File: src/mlpack/tests/ann_layer_test.cpp [hunk 2]
```diff
@@ -2931,4 +2932,33 @@ BOOST_AUTO_TEST_CASE(ConvolutionLayerPaddingTest)
...
+  module1.Parameters()(1) = 1.0;
+  module1.Parameters()(2) = 1.0;
+  module1.Parameters()(5) = 2.0;
+  module1.Parameters()(7) = 2.0;
+  module1.Parameters()(8) = 3.0;
+  module1.Parameters()(14) = 4.0;
+  module1.Reset();
+  module1.Forward(std::move(input), std::move(output));
+
+  BOOST_REQUIRE_EQUAL(arma::accu(output), 111);
```
Comment (kartikdutt18) [inline src/mlpack/tests/ann_layer_test.cpp]: Hi @zoq, I needed your help while implementing tests for this layer.
I ran the following test in python and got sum as 111 but when I run the same test using Convolution Layer I get 113. Could you help me in figuring out what I did wrong.
Thanks.

**Comment by kartikdutt18** [comment]
These needs to be rebased.

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.)

**Comment by kartikdutt18** [comment]
Lets open this. I will get back to it soon. Thanks  a lot.

**Comment by kartikdutt18** [comment]
Thanks a lot.

**Comment by Aakash-kaushik** [comment]
Hi @kartikdutt18, Was just going through this and wanted to know except the things that have changed till now(test cases, ann code refactor, etc..), is the implementation part correct ?

**Comment by kartikdutt18** [comment]
Last I remember, the test case that I wrote failed, so it probably has a bug somewhere.

**Comment by Aakash-kaushik** [comment]
> Last I remember, the test case that I wrote failed, so it probably has a bug somewhere.

All right, thanks for letting me know.
I am just going to merge it to a new branch in my repo and start working on this.

**Pull request closed without merging**